File: backend/app/services/fhir_serializer.py
from datetime import datetime, timezone
import uuid
from typing import List

from fhir.resources.patient import Patient
from fhir.resources.observation import Observation, ObservationComponent
from fhir.resources.nutritionintake import NutritionIntake, NutritionIntakeConsumedItem
from fhir.resources.condition import Condition
from fhir.resources.bundle import Bundle, BundleEntry
from fhir.resources.extension import Extension

from app.models.user import User
from app.models.glucose import GlucoseReading
from app.models.meal import MealEntry
from app.models.fasting import FastingSession

def generate_urn_uuid() -> str:
    return f"urn:uuid:{uuid.uuid4()}"

def user_to_patient(user: User) -> Patient:
    patient = Patient(
        id=str(user.id),
        identifier=[{"system": "urn:ietf:rfc:3986", "value": f"mailto:{user.email}"}],
        name=[{"use": "usual", "text": user.name}]
    )
    return patient

def glucose_to_observation(reading: GlucoseReading, patient_ref: str) -> Observation:
    obs = Observation(
        status="final",
        category=[{"coding": [{"system": "http://terminology.hl7.org/CodeSystem/observation-category", "code": "laboratory"}]}],
        code={"coding": [{"system": "http://loinc.org", "code": "15074-8", "display": "Glucose [Moles/volume] in Blood"}]},
        subject={"reference": patient_ref},
        effectiveDateTime=reading.datetime.isoformat(),
        valueQuantity={
            "value": reading.value_mgdl,
            "unit": "mg/dL",
            "system": "http://unitsofmeasure.org",
            "code": "mg/dL"
        }
    )
    
    # Add component for condition (fasting vs postprandial)
    condition_code = "1558-6" if reading.condition == "ayunas" else "1521-4"
    condition_display = "Fasting glucose" if reading.condition == "ayunas" else "Postprandial glucose"
    
    obs.component = [
        ObservationComponent(
            code={"coding": [{"system": "http://loinc.org", "code": condition_code, "display": condition_display}]}
        )
    ]
    
    return obs

def meal_to_nutrition_intake(meal: MealEntry, patient_ref: str) -> NutritionIntake:
    analysis = meal.ai_analysis or {}
    items = analysis.get("food_items", [])
    calories = analysis.get("calories_estimated", 0)
    
    ni = NutritionIntake(
        status="completed",
        code={"coding": [{"system": "http://snomed.info/sct", "code": "226379006", "display": "Food intake"}]},
        subject={"reference": patient_ref},
        occurrenceDateTime=meal.datetime.isoformat()
    )
    
    if items or calories > 0:
        desc = ", ".join(items) if items else (meal.notes or "Meal")
        ni.consumedItem = [
            NutritionIntakeConsumedItem(
                type={"text": desc},
                amount={"value": calories, "unit": "kcal", "system": "http://unitsofmeasure.org", "code": "kcal"}
            )
        ]
        
    impact = analysis.get("glycemic_impact")
    if impact:
        # Custom extension for glycemic impact
        ni.extension = [
            Extension(
                url="http://azucar.aeisoftware.com/fhir/StructureDefinition/glycemic-impact",
                valueString=impact
            )
        ]
        
    return ni

def fasting_to_observation(session: FastingSession, patient_ref: str) -> Observation:
    obs = Observation(
        status="final" if session.completed else "preliminary",
        code={"coding": [{"system": "http://snomed.info/sct", "code": "61144006", "display": "Fasting"}]},
        subject={"reference": patient_ref},
        effectivePeriod={
            "start": session.start_time.isoformat()
        },
        note=[{"text": f"Protocol: {session.protocol}"}]
    )
    if session.end_time:
        obs.effectivePeriod.end = session.end_time.isoformat()
        
    return obs

def build_patient_bundle(user: User, readings: List[GlucoseReading], meals: List[MealEntry], fastings: List[FastingSession]) -> Bundle:
    bundle = Bundle(type="collection", entry=[])
    
    # 1. Patient
    patient = user_to_patient(user)
    patient_uuid = generate_urn_uuid()
    bundle.entry.append(BundleEntry(fullUrl=patient_uuid, resource=patient))
    
    # 2. Condition (Diabetes Type 2)
    condition = Condition(
        clinicalStatus={"coding": [{"system": "http://terminology.hl7.org/CodeSystem/condition-clinical", "code": "active"}]},
        code={"coding": [{"system": "http://snomed.info/sct", "code": "44054006", "display": "Diabetes mellitus type 2"}]},
        subject={"reference": patient_uuid}
    )
    bundle.entry.append(BundleEntry(fullUrl=generate_urn_uuid(), resource=condition))
    
    # 3. Glucose Readings
    for r in readings:
        obs = glucose_to_observation(r, patient_uuid)
        bundle.entry.append(BundleEntry(fullUrl=generate_urn_uuid(), resource=obs))
        
    # 4. Meals
    for m in meals:
        ni = meal_to_nutrition_intake(m, patient_uuid)
        bundle.entry.append(BundleEntry(fullUrl=generate_urn_uuid(), resource=ni))
        
    # 5. Fastings
    for f in fastings:
        obs = fasting_to_observation(f, patient_uuid)
        bundle.entry.append(BundleEntry(fullUrl=generate_urn_uuid(), resource=obs))
        
    return bundle

File: php-backend/src/Services/FhirSerializer.php
<?php

declare(strict_types=1);

namespace Azucar\Services;

use Azucar\Support\Dt;

/**
 * Hand-built FHIR R5 JSON, ported field-for-field from fhir_serializer.py
 * (the fhir.resources dependency only added validation).
 */
final class FhirSerializer
{
    public static function urnUuid(): string
    {
        $data = random_bytes(16);
        $data[6] = chr((ord($data[6]) & 0x0f) | 0x40);
        $data[8] = chr((ord($data[8]) & 0x3f) | 0x80);
        return 'urn:uuid:' . vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
    }

    public static function userToPatient(array $user): array
    {
        return [
            'resourceType' => 'Patient',
            'id' => (string) $user['id'],
            'identifier' => [['system' => 'urn:ietf:rfc:3986', 'value' => 'mailto:' . $user['email']]],
            'name' => [['use' => 'usual', 'text' => $user['name']]],
        ];
    }

    public static function glucoseToObservation(array $reading, string $patientRef): array
    {
        $isFasting = $reading['condition'] === 'ayunas';
        return [
            'resourceType' => 'Observation',
            'status' => 'final',
            'category' => [[
                'coding' => [[
                    'system' => 'http://terminology.hl7.org/CodeSystem/observation-category',
                    'code' => 'laboratory',
                ]],
            ]],
            'code' => [
                'coding' => [[
                    'system' => 'http://loinc.org',
                    'code' => '15074-8',
                    'display' => 'Glucose [Moles/volume] in Blood',
                ]],
            ],
            'subject' => ['reference' => $patientRef],
            'effectiveDateTime' => Dt::out($reading['datetime']),
            'valueQuantity' => [
                'value' => (int) $reading['value_mgdl'],
                'unit' => 'mg/dL',
                'system' => 'http://unitsofmeasure.org',
                'code' => 'mg/dL',
            ],
            'component' => [[
                'code' => [
                    'coding' => [[
                        'system' => 'http://loinc.org',
                        'code' => $isFasting ? '1558-6' : '1521-4',
                        'display' => $isFasting ? 'Fasting glucose' : 'Postprandial glucose',
                    ]],
                ],
            ]],
        ];
    }

    public static function mealToNutritionIntake(array $meal, string $patientRef): array
    {
        $analysis = $meal['ai_analysis'] !== null ? (json_decode($meal['ai_analysis'], true) ?? []) : [];
        $items = $analysis['food_items'] ?? [];
        $calories = (int) ($analysis['calories_estimated'] ?? 0);

        $desc = $items !== [] ? implode(', ', $items) : ($meal['notes'] ?: 'Comida');

        $consumedItem = ['nutritionProduct' => ['concept' => ['text' => $desc]]];
        if ($calories > 0) {
            $consumedItem['amount'] = [
                'value' => $calories,
                'unit' => 'kcal',
                'system' => 'http://unitsofmeasure.org',
                'code' => 'kcal',
            ];
        }

        $intake = [
            'resourceType' => 'NutritionIntake',
            'status' => 'completed',
            'code' => [
                'coding' => [[
                    'system' => 'http://snomed.info/sct',
                    'code' => '226379006',
                    'display' => 'Food intake',
                ]],
            ],
            'subject' => ['reference' => $patientRef],
            'occurrenceDateTime' => Dt::out($meal['datetime']),
            'consumedItem' => [$consumedItem],
        ];

        $impact = $analysis['glycemic_impact'] ?? null;
        if ($impact) {
            $intake['extension'] = [[
                'url' => 'http://azucar.aeisoftware.com/fhir/StructureDefinition/glycemic-impact',
                'valueString' => $impact,
            ]];
        }
        return $intake;
    }

    public static function fastingToObservation(array $session, string $patientRef): array
    {
        $period = ['start' => Dt::out($session['start_time'])];
        if ($session['end_time'] !== null) {
            $period['end'] = Dt::out($session['end_time']);
        }
        return [
            'resourceType' => 'Observation',
            'status' => $session['completed'] ? 'final' : 'preliminary',
            'code' => [
                'coding' => [[
                    'system' => 'http://snomed.info/sct',
                    'code' => '61144006',
                    'display' => 'Fasting',
                ]],
            ],
            'subject' => ['reference' => $patientRef],
            'effectivePeriod' => $period,
            'note' => [['text' => 'Protocol: ' . $session['protocol']]],
        ];
    }

    /** $row must contain medication_logs columns + med_name + med_kind. */
    public static function medicationLogToStatement(array $row, string $patientRef): array
    {
        $effective = $row['marked_at'] !== null
            ? Dt::out($row['marked_at'])
            : $row['date'] . 'T00:00:00Z';

        return [
            'resourceType' => 'MedicationStatement',
            'status' => $row['status'] === 'taken' ? 'completed' : 'not-taken',
            'medication' => ['concept' => ['text' => $row['med_name']]],
            'subject' => ['reference' => $patientRef],
            'effectiveDateTime' => $effective,
            'extension' => [[
                'url' => 'http://azucar.aeisoftware.com/fhir/StructureDefinition/medication-kind',
                'valueString' => $row['med_kind'],
            ]],
        ];
    }

    public static function buildPatientBundle(array $user, array $readings, array $meals, array $fastings, array $medicationLogs): array
    {
        $entries = [];
        $patientUuid = self::urnUuid();
        $entries[] = ['fullUrl' => $patientUuid, 'resource' => self::userToPatient($user)];

        $entries[] = ['fullUrl' => self::urnUuid(), 'resource' => [
            'resourceType' => 'Condition',
            'clinicalStatus' => [
                'coding' => [[
                    'system' => 'http://terminology.hl7.org/CodeSystem/condition-clinical',
                    'code' => 'active',
                ]],
            ],
            'code' => [
                'coding' => [[
                    'system' => 'http://snomed.info/sct',
                    'code' => '44054006',
                    'display' => 'Diabetes mellitus type 2',
                ]],
            ],
            'subject' => ['reference' => $patientUuid],
        ]];

        foreach ($readings as $r) {
            $entries[] = ['fullUrl' => self::urnUuid(), 'resource' => self::glucoseToObservation($r, $patientUuid)];
        }
        foreach ($meals as $m) {
            $entries[] = ['fullUrl' => self::urnUuid(), 'resource' => self::mealToNutritionIntake($m, $patientUuid)];
        }
        foreach ($fastings as $f) {
            $entries[] = ['fullUrl' => self::urnUuid(), 'resource' => self::fastingToObservation($f, $patientUuid)];
        }
        foreach ($medicationLogs as $log) {
            $entries[] = ['fullUrl' => self::urnUuid(), 'resource' => self::medicationLogToStatement($log, $patientUuid)];
        }

        return ['resourceType' => 'Bundle', 'type' => 'collection', 'entry' => $entries];
    }
}
